common/djangoapps/third_party_auth/management/commands/saml.py

```diff
@@ -0,0 +1,151 @@
+# -*- coding: utf-8 -*-
+"""
+Management commands for third_party_auth
+"""
+import datetime
+import dateutil.parser
+from django.core.management.base import BaseCommand, CommandError
+from lxml import etree
+import requests
+from onelogin.saml2.utils import OneLogin_Saml2_Utils
+from requests.exceptions import HTTPError, SSLError
+from third_party_auth.models import SAMLConfiguration, SAMLProviderConfig, SAMLProviderData
+
+#pylint: disable=superfluous-parens,no-member
+
+
+class MetadataParseError(Exception):
+    """ An error occurred while parsing the SAML metadata from an IdP """
+    pass
+
+
+class Command(BaseCommand):
+    """ manage.py commands to manage SAML/Shibboleth SSO """
+    help = '''Configure/maintain/update SAML-based SSO'''
+
+    def handle(self, *args, **options):
+        if len(args) != 1:
+            raise CommandError("saml requires one argument: pull")
+
+        if not SAMLConfiguration.is_enabled():
+            print("Warning: SAML support is disabled via SAMLConfiguration.")
+
+        subcommand = args[0]
+
+        if subcommand == "pull":
+            self.cmd_pull()
+        else:
+            raise CommandError("Unknown argment: {}".format(subcommand))
+
+    @staticmethod
+    def tag_name(tag_name):
+        """ Get the namespaced-qualified name for an XML tag """
+        return '{urn:oasis:names:tc:SAML:2.0:metadata}' + tag_name
+
+    def cmd_pull(self):
+        """ Fetch the metadata for each provider and update the DB """
+        # First make a list of all the metadata XML URLs:
+        url_map = {}
+        for idp_slug in SAMLProviderConfig.key_values('idp_slug', flat=True):
+            config = SAMLProviderConfig.current(idp_slug)
+            if not config.enabled:
+                continue
+            url = config.metadata_source
+            if url not in url_map:
+                url_map[url] = []
+            if config.entity_id not in url_map[url]:
+                url_map[url].append(config.entity_id)
+        # Now fetch the metadata:
+        for url, entity_ids in url_map.items():
+            try:
+                print("\n→ Fetching {}".format(url))
+                if not url.lower().startswith('https'):
+                    print("→ WARNING: This URL is not secure! It should use HTTPS.")
+                try:
+                    response = requests.get(url, verify=True)
+                    response.raise_for_status()
+                except (HTTPError, SSLError):
+                    raise
```

Comment:
I suspected this was the reason for the construct.  I think a comment would be clearer, yes.  Otherwise people like me will look at the code wondering why these constructs are there.

In Python we never really know what exceptions could be thrown, so the best we can do is document as much as possible.